File: C/insertionSort.c
// INSERTION SORT- used for small size arrays

// Here you divide the array into 2 parts, sorted and unsorted. 

// Then you start picking up elements from unsorted part and place them in the proper place inside 
// the sorted part, making sure that the sorted part stays sorted. 

//  You continue this untill your array is completely sorted.

//Algorithm (ascending)
// 1: Iterate from the second element to the last element in the array
// 2: Compare current element (key) with the element before it
// 3: If key<predecessor, compare key with the elements before. 
//    ( We're moving the greater elements one position up to make space for the swapped element.


#include <math.h>
#include <stdio.h>

// Time Complexity = O(n^2) , worst case - array is sorted in reverse order, best case, array sorted.
// Auxiliary Space Needed = O(1)

void insertionSort(int arr[], int n)
{
	int i, key, j;
	for (i = 1; i < n; i++) {
		key = arr[i];
		j = i - 1;
		while (j >= 0 && arr[j] > key) 
		{
			arr[j + 1] = arr[j];
			j = j - 1;
		}
		arr[j + 1] = key;
	}
}



int main()
{
	int m;
	printf("Enter size of array : ");
	scanf("%d",&m);
	int arr[m];
	printf("\nEnter array : ");
	for(int i =0; i<m; i++){
	    scanf("%d",&arr[i]);
	}
	
	int n = sizeof(arr) / sizeof(arr[0]);

	insertionSort(arr, n);
	for ( int i = 0; i < n; i++)
		printf("%d ", arr[i]);
	printf("\n");

	return 0;
}

File: Golang/InsertionSort.go
package main

func insertionSortAlpha(arr []int) []int {
	//  Insertion sort: Works from left to right, moving the current index value to the correct position in each pass.
	//  Time Complexity: Best - O(n) Average - O(n^2) Worst - O(n^2)

	//  Declare that i, j, current are type int
	var i, j, current int

	//  Loop starting at position 1 as the value at position 0 is theoretically in a "sorted" portion of the array
	for i = 1; i < len(arr); i++ {

		//  Define the current value to sort around
		current = arr[i]

		//  Set the second index to be current - 1
		j = i - 1

		//	While loops are not implemented in Go so an alternative approach must be used
		for j >= 0 && arr[j] > current {

			//  Perform swap
			arr[j+1] = arr[j]

			//  Decrement index value of j
			j -= 1
		}

		//  Move current value into second indexer + 1 to place in correct position
		arr[j+1] = current
	}

	return arr
}

func insertionSortBeta(arr []int) []int {
	//  Insertion sort: Works from left to right, moving the current index value to the correct position in each pass.
	//  Time Complexity: Best - O(n) Average - O(n^2) Worst - O(n^2)

	//  Loop starting at position 1 as the value at position 0 is theoretically in a "sorted" portion of the array
	for i := 1; i < len(arr); i++ {

		j := i

		//	While loops are not implemented in Go so an alternative approach must be used
		for j > 0 {
			//	If the value in position j is less than the value proceeding it, swap the values
			if arr[j] < arr[j-1] {
				//	Swap the values of index j and j-1
				arr[j], arr[j-1] = arr[j-1], arr[j]
			}

			//	Decrement j
			j -= 1
		}
	}
	//	Return sorted array
	return arr
}
